File: csharp/plugins/OnePassword.cs
// 1Password Connect, as a voxgig/plugin definition.
//
// PLUGIN CODE. This file is in the VoxgigSekretoPlugins assembly, which
// the core does not reference - so nothing here is linked into an
// application whose chain names only built-in kinds. See
// docs/design/plugin-providers.md.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Net.Http;
using System.Text;

using Definition = Voxgig.Plugin.Definition;

namespace Voxgig.Sekreto.Plugins
{
    /// <summary>
    /// 1Password, through a Connect server.
    ///
    /// <para>The item titled `api.token` (titles keep their dots), in the
    /// named vault. The value is the field with purpose PASSWORD, or the
    /// field labelled `value`. A vault that cannot be found is an error -
    /// config names it, so its absence is a broken store, not a missing
    /// secret.</para>
    /// </summary>
    public class OnePasswordProvider : IProvider
    {
        private readonly string addr;
        private readonly string token;
        private readonly string vault;

        private string vaultid;

        public OnePasswordProvider(string addr = null, string token = null, string vault = null)
        {
            this.addr = addr;
            this.token = token;
            this.vault = vault;
        }

        private Dictionary<string, string> Auth()
        {
            return new Dictionary<string, string>
            {
                ["authorization"] = "Bearer " + (token ?? ""),
            };
        }

        private string ResolveVault(string useaddr)
        {
            string want = vault ?? "";
            if ("" == want)
            {
                throw new SekretoError("sekreto: onepassword: no vault");
            }

            var (status, body) = Http.FetchJson("GET", useaddr + "/v1/vaults", Auth());

            if (200 != status || !(body is List<object> listing))
            {
                throw new SekretoError("sekreto: onepassword error: " + status + ": listing vaults");
            }

            foreach (object entry in listing)
            {
                if (entry is Dictionary<string, object> found)
                {
                    object id = found.GetValueOrDefault("id");
                    if (want.Equals(id) || want.Equals(found.GetValueOrDefault("name")))
                    {
                        return Convert.ToString(id);
                    }
                }
            }

            throw new SekretoError("sekreto: onepassword: no vault named " + want);
        }

        public string Lookup(string name)
        {
            Names.CheckName(name);

            string useaddr = (addr ?? "").TrimEnd('/');
            if ("" == useaddr)
            {
                throw new SekretoError("sekreto: onepassword: no addr");
            }
            Addr.Check(useaddr);

            if (null == vaultid)
            {
                vaultid = ResolveVault(useaddr);
            }

            string filter = Url.Encode("title eq \"" + name + "\"");

            var (foundstatus, foundbody) = Http.FetchJson("GET",
                useaddr + "/v1/vaults/" + vaultid + "/items?filter=" + filter, Auth());

            if (200 != foundstatus || !(foundbody is List<object> matches))
            {
                throw new SekretoError(
                    "sekreto: onepassword error: " + foundstatus + ": finding " + name);
            }

            if (0 == matches.Count)
            {
                return null;
            }

            string itemid = Convert.ToString(
                (matches[0] as Dictionary<string, object>)?.GetValueOrDefault("id"));

            var (itemstatus, itembody) = Http.FetchJson("GET",
                useaddr + "/v1/vaults/" + vaultid + "/items/" + itemid, Auth());

            if (200 != itemstatus)
            {
                throw new SekretoError(
                    "sekreto: onepassword error: " + itemstatus + ": reading " + name);
            }

            var fields = (itembody as Dictionary<string, object>)?.GetValueOrDefault("fields")
                as List<object> ?? new List<object>();

            foreach (object entry in fields)
            {
                if (entry is Dictionary<string, object> field
                    && "PASSWORD".Equals(field.GetValueOrDefault("purpose")))
                {
                    object value = field.GetValueOrDefault("value");
                    return null == value ? null : Convert.ToString(value);
                }
            }
            foreach (object entry in fields)
            {
                if (entry is Dictionary<string, object> field
                    && "value".Equals(field.GetValueOrDefault("label")))
                {
                    object value = field.GetValueOrDefault("value");
                    return null == value ? null : Convert.ToString(value);
                }
            }

            return null;
        }

        public string Describe()
        {
            return "onepassword:" + (vault ?? "");
        }
    }

    /// <summary>The <c>onepassword</c> provider kind.</summary>
    public static class OnePassword
    {
        public static readonly Definition Plugin = Providers.ProviderPlugin(
            "onepassword", spec => new OnePasswordProvider(
                Providers.Text(spec.GetValueOrDefault("addr")),
                Providers.Text(spec.GetValueOrDefault("token")),
                Providers.Text(spec.GetValueOrDefault("vault"))));
    }
}

File: php/plugins/onepassword.php
<?php

/**
 * 1Password Connect, as a voxgig/plugin definition.
 *
 * A PLUGIN, not a built-in: it opens a socket.
 */

declare(strict_types=1);

namespace Voxgig\Sekreto\Plugins;

require_once __DIR__ . '/../src/Providers.php';
require_once __DIR__ . '/../src/Addr.php';
require_once __DIR__ . '/httpjson.php';

use Voxgig\Sekreto\Name;
use Voxgig\Sekreto\Provider;
use Voxgig\Sekreto\SekretoError;

use function Voxgig\Sekreto\checkaddr;
use function Voxgig\Sekreto\providerplugin;

/**
 * 1Password, through a Connect server.
 *
 * The item titled `api.token` (titles keep their dots), in the named
 * vault. The value is the field with purpose PASSWORD, or the field
 * labelled `value`. A vault that cannot be found is an error - config
 * names it, so its absence is a broken store, not a missing secret.
 */
class OnePasswordProvider implements Provider
{
    private ?string $vaultid = null;

    /** @param array<string, mixed> $opts */
    public function __construct(private array $opts = [])
    {
    }

    /** @return array<int, string> */
    private function auth(): array
    {
        return ['Authorization: Bearer ' . ($this->opts['token'] ?? '')];
    }

    private function resolvevault(string $addr): string
    {
        $want = $this->opts['vault'] ?? '';
        if ('' === $want || null === $want) {
            throw new SekretoError('sekreto: onepassword: no vault');
        }

        [$status, $body] = fetchjson('GET', $addr . '/v1/vaults', $this->auth());

        if (200 !== $status || !is_array($body)) {
            throw new SekretoError('sekreto: onepassword error: ' . $status . ': listing vaults');
        }

        foreach ($body as $entry) {
            if (is_array($entry)
                && ($want === ($entry['id'] ?? null) || $want === ($entry['name'] ?? null))
            ) {
                return (string) $entry['id'];
            }
        }

        throw new SekretoError('sekreto: onepassword: no vault named ' . $want);
    }

    public function lookup(string $name): ?string
    {
        Name::check($name);

        $addr = rtrim($this->opts['addr'] ?? '', '/');
        if ('' === $addr) {
            throw new SekretoError('sekreto: onepassword: no addr');
        }
        checkaddr($addr);

        if (null === $this->vaultid) {
            $this->vaultid = $this->resolvevault($addr);
        }

        $filter = rawurlencode('title eq "' . $name . '"');
        [$status, $found] = fetchjson(
            'GET',
            $addr . '/v1/vaults/' . $this->vaultid . '/items?filter=' . $filter,
            $this->auth()
        );

        if (200 !== $status || !is_array($found)) {
            throw new SekretoError(
                'sekreto: onepassword error: ' . $status . ': finding ' . $name
            );
        }

        if (0 === count($found)) {
            return null;
        }

        [$status, $item] = fetchjson(
            'GET',
            $addr . '/v1/vaults/' . $this->vaultid . '/items/' . $found[0]['id'],
            $this->auth()
        );

        if (200 !== $status) {
            throw new SekretoError(
                'sekreto: onepassword error: ' . $status . ': reading ' . $name
            );
        }

        $fields = $item['fields'] ?? [];

        foreach ($fields as $field) {
            if (is_array($field) && 'PASSWORD' === ($field['purpose'] ?? null)) {
                $value = $field['value'] ?? null;
                return null === $value ? null : (string) $value;
            }
        }
        foreach ($fields as $field) {
            if (is_array($field) && 'value' === ($field['label'] ?? null)) {
                $value = $field['value'] ?? null;
                return null === $value ? null : (string) $value;
            }
        }

        return null;
    }

    public function describe(): string
    {
        return 'onepassword:' . ($this->opts['vault'] ?? '');
    }
}

/**
 * The `onepassword` provider kind, as a voxgig/plugin definition. Pass it to
 * `Sekreto` in the `plugins` option; nothing else loads it.
 *
 * @return array<string, mixed>
 */
function onepassword(): array
{
    return providerplugin('onepassword', fn(array $spec) => new OnePasswordProvider($spec));
}
